sentry/src/main/java/io/sentry/TransactionPerformanceCollector.java

```diff
@@ -0,0 +1,84 @@
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Unmodifiable;
+
+@ApiStatus.Internal
+public final class TransactionPerformanceCollector {
+  private static final long TRANSACTION_COLLECTION_INTERVAL_MILLIS = 100;
+  private static final long TRANSACTION_COLLECTION_TIMEOUT_MILLIS = 30000;
+  private @NotNull Timer timer = new Timer();
+  private final @NotNull Map<String, ArrayList<MemoryCollectionData>> memoryMap =
+      new ConcurrentHashMap<>();
+  private final @NotNull IMemoryCollector memoryCollector;
+  private final @NotNull SentryOptions options;
+  private final @NotNull AtomicBoolean isStarted = new AtomicBoolean(false);
+
+  public TransactionPerformanceCollector(final @NotNull SentryOptions options) {
+    this.options = Objects.requireNonNull(options, "The options object is required.");
+    this.memoryCollector = options.getMemoryCollector();
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public void start(final @NotNull ITransaction transaction) {
+    if (memoryCollector instanceof NoOpMemoryCollector) {
+      options
+          .getLogger()
+          .log(
+              SentryLevel.INFO,
+              "Memory collector is a NoOpMemoryCollector. Memory stats will not be captured during transactions.");
+      return;
+    }
+    if (!memoryMap.containsKey(transaction.getEventId().toString())) {
+      memoryMap.put(transaction.getEventId().toString(), new ArrayList<>());
+      options
+          .getExecutorService()
+          .schedule(
+              () -> {
+                ArrayList<MemoryCollectionData> memoryCollectionData =
+                    (ArrayList<MemoryCollectionData>) stop(transaction);
+                memoryMap.put(transaction.getEventId().toString(), memoryCollectionData);
+              },
+              TRANSACTION_COLLECTION_TIMEOUT_MILLIS);
+    }
+    if (!isStarted.getAndSet(true)) {
+      timer.scheduleAtFixedRate(
+          new TimerTask() {
+            @Override
+            public void run() {
+              MemoryCollectionData memoryData = memoryCollector.collect();
+              if (memoryData != null) {
+                synchronized (memoryMap) {
+                  for (ArrayList<MemoryCollectionData> list : memoryMap.values()) {
+                    list.add(memoryData);
+                  }
+                }
+              }
+            }
+          },
+          0,
+          TRANSACTION_COLLECTION_INTERVAL_MILLIS);
+    }
+  }
+
+  public @Unmodifiable List<MemoryCollectionData> stop(final @NotNull ITransaction transaction) {
+    synchronized (memoryMap) {
+      List<MemoryCollectionData> memoryData = memoryMap.remove(transaction.getEventId().toString());
+      if (memoryMap.isEmpty() && isStarted.get()) {
```

Comment:
`m`: You can utilize `compareAndSet()` to read and update the AtomicBoolean at the same time
```suggestion
      if (memoryMap.isEmpty() && isStarted.compareAndSet(true, false)) {
```